### ISSUE TITLE
[API Proposal]: Break out [Uri Class (System)](https://learn.microsoft.com/en-us/dotnet/api/system.uri?view=net-7.0) query parameters as a property

### ISSUE THREAD

**Issue by WillPittenger** [dotnet/runtime#104986]
### Background and motivation

Currently, when you have a [Uri Class (System)](https://learn.microsoft.com/en-us/dotnet/api/system.uri?view=net-7.0) instance, it doesn't break out the query parameters as separate values.  You have to go through a completely different API that's hard to find.  The [Query](https://learn.microsoft.com/en-us/dotnet/api/system.uri.query?view=net-7.0#system-uri-query) property simply returns a string.  I have no idea how you'd allow users to change parameters.  The existing class doesn't seem to be marked Immutable, but all properties are read-only.

### API Proposal

```csharp
namespace System.Uri;

public class Uri : System.Runtime.Serialization.ISerializable
{
    .
    .
    .

    public System.Collections.Generic.IReadOnlyDictionary<string, Object> QueryParameters
    {
        get;
    }

    .
    .
    .

    public Uri MakeNewWithChangedParameters(System.Collections.Generic.IReadOnlyMap<string, string> newOrChangedParameters = null, params string[] deletedParameters = null);

    .
    .
    .
}
```
You might want to add new constructors too.  I didn't show those.

### API Usage

```csharp
System.Uri uri = new System.Uri("https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&projects=&template=02_api_proposal.yml&title=%5BAPI+Proposal%5D%3A+");

foreach(string s in uri.QueryParameters.Keys)
    System.Diagnositics.Debug.Log($"{s}: {uri.QueryParameters[s]}");

System.Uri uri2 = uri.MakeNewWithChangedParameters(new System.Collections.Generic.Dictionary<string, Object> {"assignee" = "t", "junk" = 5}, "projects");
```

**Comment by julealgon** [comment]
Watch out for the use of a dictionary there. Depending on how this is represented, that would not work.

Keep in mind that having many entries in the queryString area with the same key is completely valid, and is usually the canonical form of passing multiple values for an array property:
```
"http://myServer/endpoint?myArray=1&somethingElse=anotherValue&myArray=2"
```

Either the abstraction should allow for the duplicates so a client can parse them, or it should have a facility to parse the value directly, i.e.:
```
var myArray = uri.QueryParameters["myArray"]; // returns either a List or IEnumerable here
```

In a way, this is similar to how Http Headers work, so maybe it could also use similar abstractions like [`StringValues`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.stringvalues?view=net-8.0) here to represent the value for each key.

**Comment by WillPittenger** [comment]
So something more like `System.Collections.Generic.IReadOnlyDictionary<string, string[]>`?

**Comment by rjgotten** [comment]
> So something more like `System.Collections.Generic.IReadOnlyDictionary<string, string[]>`?

No. At best `IReadOnlyDictionary<string?,string[]>` Because you can have keyless values.
Actually, the _actual_ best is just what it already is: `string`

Because query strings _do not have any universal agreed upon format mandated by official spec_.
All of it is done 'by convention' which can shift from application to application.

**Comment by wvpm** [comment]
> Because you can have keyless values - e.g. ?a

Should the key be `""` is this case?
That makes more sense than null. Otherwise `""` and null might be 2 distinct keys.

**Comment by rjgotten** [comment]
> > Because you can have keyless values - e.g. ?a
> 
> Should the key be `""`/`string.Empty` is this case? That makes more sense than null. Otherwise `""` and null might be 2 distinct keys.

That'd be pretty much a necessity as dictionaries don't take `null` keys.

**Comment by MihaZupan** [comment]
Closing this in favour of discussions in #18874/#32606 since the concepts overlap.
